### PR TITLE
Da BIgg refactor

### PR DESCRIPTION
**Pull request by ss2165**

### CODE REVIEW

File: src/types/new_type_row.rs [hunk 1]
```diff
@@ -88,6 +90,23 @@ impl<T: TypeClass> TypeRow<T> {
...
+    /// given the rows of each tuple
+    pub fn predicate_variants_row(variant_rows: impl IntoIterator<Item = TypeRow<T>>) -> Self {
+        variant_rows
+            .into_iter()
+            .map(Type::new_tuple)
+            .collect_vec()
+            .into()
+    }
+
+    /// Returns the smallest [TypeTag] that contains all elements of the row
```
Comment (acl-cqc) [inline src/types/new_type_row.rs:104]: or, erm, was it `actual_tag`?? (I do not insist, having a different name is reasonable)

File: src/types/new_type_row.rs [hunk 2]
```diff
@@ -88,6 +90,23 @@ impl<T: TypeClass> TypeRow<T> {
                 .collect::<Vec<Type<T2>>>(),
         )
     }
+
+    /// Return the type row of variants required to define a Sum of Tuples type
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/types/new_type_row.rs:106]: yes done, but for Any

[the remark this answers is not in the document]
Comment (acl-cqc) [reply src/types/new_type_row.rs:106]: ok here we go

File: src/types/type_enum.rs [hunk 3]
```diff
@@ -51,6 +62,22 @@ impl<T: TypeClass> Type<T> {
...
+impl Type<AnyLeaf> {
+    /// New Sum of Tuple types, used as predicates in branching.
+    /// Tuple rows are defined in order by input rows.
+    pub fn new_predicate<V>(variant_rows: impl IntoIterator<Item = V>) -> Self
+    where
+        V: IntoIterator<Item = Type<AnyLeaf>>,
+    {
+        Self::new_sum(predicate_variants_row(variant_rows))
+    }
+    /// New simple predicate with empty Tuple variants
```
Comment (acl-cqc) [inline src/types/type_enum.rs:74]: nit: newline in wrong place (should be before this comment not after)

File: src/types.rs [hunk 4]
```diff
@@ -530,7 +529,7 @@ mod test {
     use super::*;
     #[test]
     fn construct() {
-        let t: Type<CopyableLeaf> = Type::new_tuple([
+        let t: Type<CopyableLeaf> = Type::new_tuple(vec![
```
Comment (acl-cqc) [inline src/types.rs:532]: think you can undo this

**Review by acl-cqc — APPROVED** [review]
Ok, squash-merge into `types/redux` and we're good to go

**Pull request merged**